src/RouterUtil.js

```diff
@@ -0,0 +1,104 @@
+/*
+ * routeStack : [userRoute, userRoute2, ...]
+ * processStack: [
+ * { type: push| pop, data: userRoute },
+ * { type: push| pop, data: userRoute2 },
+ * ...
+ * ]
+ */
+
+export default {
+  init: (routes) => {
+    return {
+      routeStack: routes,
+      processStack: []
+    };
+  },
+
+  replace: ({routeConfig, data, options, key}) => {
+    let config = {...routeConfig};
+
+    // do not push keys twice
+    if (key == null ||
+      config.processStack.filter((el) => el.key === key).length === 0) {
+      config.processStack.push({
+        type: 'replace',
+        data,
+        options,
+        key
+      });
+    }
+
+    return config;
+  },
+
+  reset: ({routeConfig, data, options, key}) => {
+    let config = {...routeConfig};
+
+    // do not push keys twice
+    if (key == null ||
+      config.processStack.filter((el) => el.key === key).length === 0) {
+      config.processStack.push({
+        type: 'reset',
+        data,
+        options,
+        key
+      });
+    }
+
+    return config;
+  },
+  push: ({routeConfig, data, options, key}) => {
```

Comment:
Maybe we should call it `route` instead of `data` ? I think that might make more sense.